TESTING.md

```diff
@@ -0,0 +1,274 @@
+
+```ts
+const {getByTestId} = render(
+  <SomeComponent data-testid="container">
+    <div data-testid="test">Test</div>
+  </SomeComponent>
+);
+
+const component = await getByTestId('container');
+const child = await getByTestId('test');
+
+expect(component).toContainElement(child);
+```
+
+The first test has the DOM structure encoded in it. The second
+
+This example is more flexible. The component can be refactored with DOM elements added or removed
+and the test will still pass as long as the `div` is a child of the component somewhere (the
+specification is met). Also this example will have a more useful failure message since
+`.toContainElement` has the context that it is expecting an element in another element vs a match of
+a string.
+
+## Functional tests
+
+Canvas Kit uses [Cypress][https://cypress.io] for browser-based behavior testing (additional info:
+[Why Cypress?](./cypress/WHY_CYPRESS.md)). Functional tests ensure the code meets functional
+specifications from a user's perspective. Specifications can come from many different stakeholders
+including product management, user research and accessibility. A functional test should read in the
+Given/When/Then format and fits in well as acceptance criteria.
+
+Example of a usability specification: "Given the modal example, when a user clicks the 'Delete Item'
+button, then the modal opens".
+
+Example of an accessibility specification: "Given the modal example, when a user clicks the 'Delete
+Item' button, then the focus should be transferred to the first focusable element in the dialog"
+
+The BDD-style (behavior-driven development) syntax lends well to the nesting of actions performed on
+a example (also called "story" borrowed from Storybook nomenclature). The wording used inside the
+BDD blocks should read in English and be understandable by non-developers. The implementation should
+be as readable as possible by non-developers for trust and verification. The intent of the wording
+of the BDD-style syntax is to parse into an outline form to go into documentation as a verification
+of what specifications are covered.
+
+Example:
+
+```ts
+describe('Given the Default Modal is rendered', () => {
+  beforeEach(() => {
+    // render the default modal example
+  });
+
+  context('when the target button is clicked', () => {
+    beforeEach(() => {
+      // click the target button
+    });
+
+    it('should open the modal', () => {
+      // expect modal to be open
+    });
+  });
+});
+```
+
+The outline would look like:
+
+- Given the Default Modal is rendered
+  - when the target button is clicked
+    - then it should open the modal
+
+Tests should rely on aria attributes as much as possible for selecting and asserting. Test IDs
+should be used only to disambiguate DOM elements.
+
+Read [Intro to Cypress](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html)
+for a good idea of how Cypress works. The core concept to Cypress is enqueued command chains that
```

Comment:
```suggestion
for a good idea on how Cypress works. The core concept to Cypress is enqueued command chains that
```

Reply:
"for an intro to how Cypress works."?

Reply:
Hmm. Maybe that's redundant if the link text is "Intro to Cypress"